codePost_lib.py

```diff
@@ -0,0 +1,695 @@
+
+        # CASE 1: No existing submission => create a new submission
+
+        return post_submission(
+            api_key=api_key,
+            assignment_id=assignment_id,
+            students=students,
+            files=files
+        )
+
+    # There is at least one (maybe more) existing submissions
+
+    # First check the modes to determine whether to proceed.
+    if not mode["updateIfExists"]:
+        raise UploadError(
+            """
+            At least one submission already exists, and 'updateIfExists' is false,
+            so interrupting upload.
+            """)
+
+    # Check whether any of the existing submissions are claimed.
+    if not mode["updateIfClaimed"] and _submission_list_is_unclaimed(existing_submissions):
+        raise UploadError(
+            """
+            At least one submission has already been claimed by a grader, and
+            'updateIfClaimed' is false, so interrupting upload.
+            """)
+
+    # Check whether students will need an update.
+    if not mode["resolveStudents"]:
+        if set(students) != set(existing_submissions[0]["students"]) or len(existing_submissions) > 1:
+            raise UploadError(
+                """
+                There are {} existing submission(s) with a different subset of
+                students than those requested. Since 'resolveStudents' is false,
+                interrupting upload.
+                - Requested students: {}
+                - Existing students (on first existing submission): {}
+                """.format(
+                    len(existing_submissions),
+                    students,
+                    set(existing_submissions[0]["students"])
+                ))
+
+    if len(existing_submissions) > 1:
+
+        # CASE 2: Remove the students that we need to assign to the uploaded submission
+        # from their existing submissions
+
+        for submission in existing_submissions:
+            changed_submission = remove_students_from_submission(
+                api_key=api_key,
+                submission_info=submission,
+                students_to_remove=students
+            )
+
+            if mode["deleteAffectedSubmissions"]:
+                delete_submission(
+                    api_key=api_key,
+                    submission_id=changed_submission["id"]
+                )
+
+        return post_submission(
+            api_key=api_key,
+            assignment_id=assignment_id,
+            students=students,
+            files=files
+        )
+
+    # CASE 3: There is exactly one submission.
+    submission = existing_submissions[0]
+    submission_id = submission["id"]
+
+    # Update the submission students to make sure it is what was specified (if we needed
+    # to make this change, and it was forbidden, this would already have been caught).
+    set_submission_students(
+        api_key=api_key,
+        submission_id=submission_id,
+        students=students
+    )
+
+    if mode["addFiles"]:
+
```

Comment:
What about the case for which `mode["updateExistingFiles"]` is `True` but `mode["addFiles"]` is False? This doesn't map to the old upload modes, but is still a valid mode for a caller to require. 

In this situation, we'd still want to call `diffscan_submission`